Use fetch + async/await instead of axios

temsa opened the pull request.

bmullan91 commented on src/index.js:
```diff
@@ -2,11 +2,12 @@ require('@babel/polyfill')
-    console.error(err)
-    process.exit(1)
-  })
+;(async () => {
```
Same comment from `scripts/lighthouse.js`

temsa replied:
@bmullan91 same linter issue ( at least ), I originally put things like that because I had an error (might be a linter one already at that time), because an `await` should be inside an `async function`

bmullan91 commented on src/server/index.js:
```diff
@@ -39,29 +39,25 @@ app.get('/api/stories', (request, response) => {
 
 function sanitizeItemContent (items) {
-  return items.map(({content, ...rest}) => {
+  return items.map(({ content, ...rest }) => {
```
👍

bmullan91 commented on src/server/index.js:
```diff
@@ -109,47 +105,39 @@ function graphQLResponse (filter, page, response) {
     })
-    .catch(error => {
-      error.statusCode = error.statusCode || 500
+    const result = await res.json()
```
We need to re-add the error handling - axios always threw an error if the status code was > 200. Where as the fetch api doesn't.

Same idea from my previous comment:

```js
const response = await fetch(url)

if (response.ok) {
  // all good
} else {
 // handle error - use existing logic
}
```

temsa replied:
:ok_hand: that's right, though ideally we should check for emptiness of the response or invalid content too, a reverse proxy can for example provide a 200 with an html instead of the json and it should be an error

bmullan91 replied:
You're right - we're doing the bare minimum here and getting away with it 😅

bmullan91 commented on src/server/index.js:
```diff
@@ -109,47 +105,39 @@ function graphQLResponse (filter, page, response) {
+  const port = 3000
+  const listener = new Promise((resolve, reject) => app.listen(port, err => err ? reject(err) : resolve(app)))
+  const server = await listener
+  return { app, server, port }
```
Nice one 👍

temsa commented on src/app/helpers/stories.js:
```diff
@@ -1,18 +1,23 @@
-  return axios.get(`${HOST}/api/stories?${params}`)
+  const res = await fetch(`${HOST}/api/stories?${params}`)
+  const data = await res.json()
+  if (data.message && data.stack && data.statusCode) {
```
This implementation would eradicate the original `data.stack` to replace it with a new one coming from the `new Error`. It has its pros and cons, but should it be part of the data, we should not drop it and at least put in somewhere in the resulting error

bmullan91 replied:
Yea we can re-use the true error stack - accidentally left it out in the pseudo code.

temsa commented on src/app/routes.js:
```diff
@@ -5,7 +5,7 @@ module.exports = [
     exact: true,
     component: require('./pages/LoadingPage'),
-    fetcher: () => Promise.resolve({ data: {} })
+    fetcher: async () => ({ data: {} })
```
Why did `Promise.resolve` was, then ?

bmullan91 replied:
The fetcher should return a promise.

temsa replied:
This version feels to me more like other routes, and has the same result, don't you think ? DO you really want to put back the Promise here ?

temsa commented on src/client/js/app-shell.js:
```diff
@@ -1,3 +1,4 @@
+require('babel-polyfill')
```
Actually it seems to make the bundle bigger, and I find it quite a deal breaker for something that should be smaller in principle ( fetch vs. axios ), I'll add a `.browserlistrc` to check if not supporting every browser on earth helps with that

bmullan91 replied:
@temsa do you know how much bigger? The browserlist is definitely something we need to do - although probably warrants a separate ticket as it will require some investigation.

temsa replied:
It was ~90kb bigger. Finally replacing `babel-polyfill` for the browser by [transform-runtime](http://babeljs.io/docs/plugins/transform-runtime/) allowed to only keep what was necessary (thanks to tree shaking too). It is still not smaller than the original `axios` version , but at least it is only `8kb` bigger.
This is still frustrating and can probably be improved again, at least the tools are in place to do it (including a new `analyze` script to check the built bundle`

temsa commented on scripts/lighthouse.js:
```diff
@@ -18,16 +19,16 @@ const opts = {
-    console.error(err)
-    process.exit(1)
-  })
+;(async () => {
```
@bmullan91 it seems to bee needed, at least by the linter, if I remove it, I obtain that error (and I've configured standard to support  browser + es6 + node):
```console
husky > npm run -s precommit (node v10.1.0)

standard: Use JavaScript Standard Style (https://standardjs.com)
  /home/temsa/dev/nfm/react-pwa/scripts/lighthouse.js:22:28: Parsing error: await is a reserved word

  20 | 
  21 | try {
> 22 |   const { server, port } = await init()
     |                            ^
  23 |   console.log(`\n\n Server started on port `, port)
  24 | 
  25 |   const results = await launchChromeAndRunLighthouse(`http://127.0.0.1:${port}/`, opts)

```

temsa commented:
@bmullan91 I believe I fixed all the issues you have found (notably error handling), plus the ones I have found:
- the resulting bundle was too big ( ~80kb bigger than previous implementation )
- smarter and smaller polyfilling using babel's `transform-runtime` : original implementation did not have polyfilling for frontend, this one does but only takes what it needs
- missing tree shaking for smaller bundle: necessitated es6 imports to work well
- missing `.browserlistrc`: for babel, etc. to only produce code for the queried browsers, plus a npm script to see the actual supported browser list according to the query
- nothing to analyze the bundle: a small script adding no dependency ( `npx` powered )
- nothing to debug easily the transpiled server code : added the sourcemap so it can be `inspect`ed easily
- all babel ecosystem version were not the same, updated them to the latest one
- `upath` dependency was missing should I run a `yarn` or `npm install`, maybe due to some dependency update in this PR ( I did not check back `master` tbh)

temsa commented:
We will go with @ShogunPanda full rewrite #49 , the PR can have a few commits interesting to pick-up, but can be closed
